File: CamelMatch.java
class CamelMatch {
    //time: O((m+n)*k) where m = avg length of a word, n=avg length of pattern, k= avg length of input list
    //space: O(1)
    //did it run on leetcode: yes
    public List<Boolean> camelMatch(String[] queries, String pattern) {
        //list for the output
        List<Boolean> res = new ArrayList<>();
        //looping through all the words in list
        for(String query: queries){
            //initializing pointer for pattern
            int j=0;
            //initializing flag to track if current word is in the pattern
            boolean flag = false;
            //looping through current word
            for(int i=0; i<query.length(); i++){
                //fetching current character
                Character c = query.charAt(i);
                //if pattern pointer is within bound and current character of word and pattern matches
                if(j<pattern.length() && c == pattern.charAt(j)){
                    //update pattern pointer
                    j++;
                    //if pattern pointer at the end
                    if(j==pattern.length()){
                        //update flag as pattern found
                        flag = true;
                    }  
                }//if any uppercase character found not in pattern
                else if(Character.isUpperCase(c)){
                    //mark flag as pattern mismatch
                        flag = false;
                        break;
                }
            }
            //add to result
            res.add(flag);
        }
        return res;
    }
}

File: camelMatch.cpp
// Time complexity: O(m+n)k where m is the length of string, n is the number of queries and k is the pattern len
// Space complexity: O(1)

/*
The solution is true when:
Pattern matches and we reach the end of the pattern 
provided we don't see any new Capital letters in the query 
*/

class Solution {
public:
    vector<bool> camelMatch(vector<string>& queries, string pattern) {
        if(queries.size() == 0)
            return vector<bool>();

        vector<bool> result;
        for(string query: queries) {
            int p = 0;
            bool match = false;
            for(int q = 0; q<query.length(); q++) {
                if(p < pattern.length() && pattern[p] == query[q]) {
                    p++;
                    if(p == pattern.length())
                        match = true;
                }
                else if(query[q] >= 'A' && query[q] <= 'Z') {
                    match = false;
                    break;
                }
            }
            result.push_back(match);
        }
        return result;
    }
};